common/test/unit/com/thoughtworks/go/domain/PipelineLabelTest.java

```diff
@@ -129,11 +161,126 @@ public void shouldCreateLabelIfTemplateIsProvided() {
+
+    @BeforeClass
+    public static void setup() {
+        MATERIAL_REVISIONS.put(new CaseInsensitiveString("svnRepo.verynice"), SVN_REVISION);
+        MATERIAL_REVISIONS.put(new CaseInsensitiveString("svn"), SVN_REVISION);
+        MATERIAL_REVISIONS.put(new CaseInsensitiveString("git"), GIT_REVISION);
+    }
+
+    @Test
+    public void canMatchMaterialName() throws Exception {
+        final String[][] expectedGroups = { { "git" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${git}", expectedGroups);
+        assertThat(res, is("release-" + GIT_REVISION));
+    }
+
+    @Test
+    public void canMatchMaterialNameWithTrial() throws Exception {
+        final String[][] expectedGroups = { { "git" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${git}.alpha.0", expectedGroups);
+        assertThat(res, is("release-" + GIT_REVISION + ".alpha.0"));
+    }
+
+    @Test
+    public void canHandleWrongMaterialName() throws Exception {
+        final String[][] expectedGroups = { { "gitUnused" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${gitUnused}", expectedGroups);
+        assertThat(res, is("release-${gitUnused}"));
+    }
+
+    @Test
+    public void canMatchWithoutTruncation() throws Exception {
+        final String[][] expectedGroups = { { "svnRepo.verynice" }, { "git" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${svnRepo.verynice}-${git}", expectedGroups);
+        assertThat(res, is("release-" + SVN_REVISION + "-" + GIT_REVISION));
+    }
+
+    @Test
+    public void canMatchWithOneGitTruncation() throws Exception {
+        final String[][] expectedGroups = { { "git", "7" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${git[:7]}", expectedGroups);
+        assertThat(res, is("release-" + GIT_REVISION.substring(0, 7)));
+    }
+
+    @Test
+    public void canMatchWithOneGitTruncationTooLongToTruncate() throws Exception {
+        final String[][] expectedGroups = { { "git", "9999" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${git[:9999]}", expectedGroups);
+        assertThat(res, is("release-" + GIT_REVISION));
+    }
+
+    @Test
+    public void canMatchWithOneGitTruncationAlmostTruncated() throws Exception {
+        final String[][] expectedGroups = { { "git", GIT_REV_LENGTH + "" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${git[:" + GIT_REV_LENGTH + "]}", expectedGroups);
+        assertThat(res, is("release-" + GIT_REVISION));
+    }
+
+    @Test
+    public void canMatchWithOneGitTruncationByOneChar() throws Exception {
+        final int size = GIT_REV_LENGTH - 1;
+        final String[][] expectedGroups = { { "git", size + "" } };
+        String res = assertLabelGroupsMatchingAndReplace("release-${git[:" + size + "]}", expectedGroups);
+        assertThat(res, is("release-" + GIT_REVISION.substring(0, GIT_REV_LENGTH - 1)));
```

Comment:
Maybe `canMatchWithOneGitTruncationByOneChar` is not self-explaining, but I currently have not a better idea how to do so.

I aimed for a boundary test: truncating exactly one char.